packages/java/endpoint/src/main/java/com/vaadin/hilla/route/RouteUnifyingIndexHtmlRequestListener.java

```diff
@@ -128,28 +128,26 @@ public void modifyIndexHtmlResponse(IndexHtmlResponse response) {
     }
 
     protected Map<String, AvailableViewInfo> collectClientViews(
-            Predicate<? super String> isUserInRole,
-            boolean isUserAuthenticated) {
+            VaadinRequest request) {
+
         if (!deploymentConfiguration.isProductionMode()) {
             clientRouteRegistry.loadLatestDevModeFileRoutesJsonIfNeeded(
                     deploymentConfiguration);
         }
 
-        return clientRouteRegistry.getAllRoutes().values().stream()
-                .filter(config -> config.getRouteParameters() == null
-                        || config.getRouteParameters().isEmpty()
-                        || config.getRouteParameters().values().stream()
-                                .noneMatch(
-                                        param -> param == RouteParamType.REQUIRED))
-                .filter(config -> routeUtil.isRouteAllowed(isUserInRole,
-                        isUserAuthenticated, config))
-                .map(config -> new AvailableViewInfo(config.getTitle(),
-                        config.getRolesAllowed(), config.isLoginRequired(),
-                        config.getRoute(), config.isLazy(),
-                        config.isAutoRegistered(), config.menu(),
-                        Collections.emptyList(), config.getRouteParameters()))
-                .collect(Collectors.toMap(AvailableViewInfo::route,
-                        Function.identity()));
+        return MenuRegistry
```

Comment:
This part was trying to optimize loading of the latest changes in the client-side menus (in dev mode), and now that they are being collected by MenuRegistry, it seems a bit pointless. `ClientRouteRegistry` was the only source of truth about the routes and was preventing concurrency by having all methods `synchronized` due to having some  states: 

* `registeredRoutes` (originally for optimizations)
* `hasMainLayout` (which can be determined only when traversing the json hierarchy of menus). 
 
The optimizations during the Dev mode is not very critical, but the `hasMainLayout` is used for handling this logic, which I'm not sure how is it handled anymore, maybe on Flow side?

Reply:
The MenuRegistry always recreates the objects from the actual file. We could have caching there or here if it's necessary, but at the moment it is always generated and thus only a change to the file while reading could mess things up.

hasMainLayout does not have a method in flow as there is no mainLayout there.
Looking at it now it is populated for non prod mode running the `loadLatestDevModeFileRoutesJsonIfNeeded` method. So this is something that is not fully correct.

 What is the requirement for a mainLayout? Is it just no route but has childern and is not a child of anyone? As that we could probably check from the results?

Reply:
> it is populated for non prod mode

The `hasMainLayout` is initialized in `registerAndRecurseChildren` which is called in `public synchronized void registerClientRoutes`, so it is not only for Dev mode, the `RouteUnifyingServiceInitListener` also calls it in the Prod mode.

> What is the requirement for a mainLayout?

For generating an automatic menu, the requirement was defined as: Server side routes should only be populated to the `window.Vaadin.views` if a main layout exists (supposedly the menu is rendered there). That logic was checked [here](https://github.com/vaadin/hilla/blob/main/packages/java/endpoint/src/main/java/com/vaadin/hilla/route/RouteUnifyingIndexHtmlRequestListener.java#L169-L175).

> Is it just no route but has childern and is not a child of anyone?

A layout is defined as an entry in the `file-routes.json` that has:
- has no parent (for distinguishing it from intermediary layouts), and
- `"route": ""`, and
- `"children": []` or `"children": [some entries]` (`"children"` should not be undefined)

"having no parents" was basically one of the motivations for not loading and traversing the `file-routes.json` and trying to infer it while reading that file only when it is necessary.

Reply:
@Lodin 
According to the changes in this PR, what is sent the browser is collected by `MenuRegistry`, but the other parts of Hilla will still be relying on the loaded cache of `ClientRouteRegistry` (it means codes calling `clientRouteRegistry::getAllRoutes` are getting whatever is already collected in the `menuRegistry` map). I still believe there should be only one source of truth since these two can easily diverge in the near future and get out of sync.